gem/optimise.py

```diff
@@ -413,3 +475,555 @@ def aggressive_unroll(expression):
+    elif isinstance(a, Sum) and set(a.free_indices) & self.index_set:
+        return Sum(self(Product(a.children[0], b)),
+                   self(Product(a.children[1], b)))
+    else:
+        return Product(a, b)
+
+
+# TODO: arguablly should move this inside LoopOptimiser
+def expand_products(node, indices):
+    """
+    Expand products recursively if free indices of the node contains index
+    from :param indices
+    e.g (a+(b+c)d)e = ae + bde + cde
+    :param node: gem expression
+    :param indices: tuple of indices
+    :return: gem expression with products expanded
+    """
+    mapper = Memoizer(_expand_products)
+    mapper.index_set = set(indices)
+    return mapper(node)
+
+
+def collect_terms(node, node_type):
+    """Recursively collect all children into a list from :param:`node`
+    and its children of class :param:`node_type`.
+
+    :param node: root of expression
+    :param node_type: class of node (e.g. Sum or Product)
+    :return: list of all terms
+    """
+    terms = []
+    queue = [node]
+    while queue:
+        child = queue.pop()
+        if isinstance(child, node_type):
+            queue.extend(child.children)
+        else:
+            terms.append(child)
+    return tuple(terms)
+
+
+@singledispatch
+def _hoist_indexsum(node, self):
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_hoist_indexsum.register(Node)(reuse_if_untouched)
+
+
+@_hoist_indexsum.register(IndexSum)
+def _hoist_indexsum_indexsum(node, self):
+    child = self(node.children[0])
+    if isinstance(child, Product):
+        mi_set = set(node.multiindex)
+        factors = collect_terms(child, Product)
+        hoisted_factors = []
+        remaining_factors = []
+        for factor in factors:
+            if set(factor.free_indices) & mi_set:
+                remaining_factors.append(factor)
+            else:
+                hoisted_factors.append(factor)
+        if hoisted_factors:
+            new_indexsum = IndexSum(reduce(Product, remaining_factors, one), node.multiindex)
+            return Product(reduce(Product, hoisted_factors, one), new_indexsum)
+    return node
+
+
+def hoist_indexsum(node):
+    """
+    lifting factors invariant in the contraction indices out of an IndexSum
+    e.g. IndexSum(A*B, (i,)) = A * IndexSum(B), if A is independent of i
+    :param node:
+    :return:
+    """
+    mapper = Memoizer(_hoist_indexsum)
+    return mapper(node)
+
+
+def decide_key(factor, arg_ind_flat, arg_ind_set):
+    """
+    Helper function to decide the appropriate key of a factor
+    :param factor: gem node
+    :param arg_ind_flat: set of argument indices
+    """
+    fi = factor.free_indices
+    if not fi:
+        return 'const'
+    else:
+        ind_set = set(fi) & arg_ind_set
+        if len(ind_set) > 1:
+            return tuple(i for i in arg_ind_flat if i in ind_set)
+        elif len(ind_set) == 0:
+            return 'other'
+        else:
+            return ind_set.pop()
```

Comment:
This line is also the point where you emit a single index object instead of a 1-tuple of indices (inconsistency).